trio/_core/_io_windows.py

```diff
@@ -109,6 +110,65 @@ def _handle(obj):
         return obj
 
 
+async def WaitForSingleObject(handle):
+    """Async and cancellable variant of kernel32.WaitForSingleObject().
+
+    Args:
+      handle: A win32 handle.
+
+    """
+    # Quick check; we might not even need to spawn a thread. The zero
+    # means a zero timeout; this call never blocks. We also exit here
+    # if the handle is already closed for some reason.
+    retcode = kernel32.WaitForSingleObject(handle, 0)
+    if retcode != ErrorCodes.WAIT_TIMEOUT:
+        return
+
+    # :'(  avoid circular imports
+    from .._threads import run_sync_in_worker_thread
+
+    class StubLimiter:
+        def release_on_behalf_of(self, x):
+            pass
+
+        async def acquire_on_behalf_of(self, x):
+            pass
+
+    # Wait for a thread that waits for two handles: the handle plus a handle
+    # that we can use to cancel the thread.
+    cancel_handle = kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)
+    try:
+        await run_sync_in_worker_thread(
+            WaitForMultipleObjects_sync,
+            handle,
+            cancel_handle,
+            cancellable=True,
+            limiter=StubLimiter(),
+        )
+    finally:
+        # Clean up our cancel handle. In case we get here because this task was
+        # cancelled, we also want to set the cancel_handle to stop the thread.
+        kernel32.SetEvent(cancel_handle)
+        kernel32.CloseHandle(cancel_handle)
+
+
+def WaitForMultipleObjects_sync(*handles):
+    """Wait for any of the given Windows handles to be signaled.
+    
+    """
+    n = len(handles)
+    handle_arr = ffi.new("HANDLE[{}]".format(n))
+    for i in range(n):
+        handle_arr[i] = handles[i]
+    timeout = 1000 * 60 * 60 * 24  # todo: use INF here, whatever that is, and ditch the while
+    while True:
+        retcode = kernel32.WaitForMultipleObjects(
+            n, handle_arr, False, timeout
+        )
+        if retcode != ErrorCodes.WAIT_TIMEOUT:
+            break
```

Comment:
The docs say that `WaitForMultipleObjects` can return `WAIT_FAILED`, and that in this case we should check `GetLastError`. I can't think of why this would ever fail, but probably best to check anyway to be safe. Fortunately `_windows_cffi.py` already has the machinery you need; just do something like:

```python
if retcode == WAIT_FAILED:
    _windows_cffi.raise_winerror()
```